File: src/nam.rs
use std::fmt::{Display, Formatter};
use std::time::Instant;

use fastrand::Rng;
use log::Level::Trace;
use log::trace;

use crate::chainer::Anchor;
use crate::chainer::Chainer;
use crate::details::NamDetails;
use crate::index::StrobemerIndex;
use crate::io::fasta::RefSequence;
use crate::mcsstrategy::McsStrategy;
use crate::read::Read;
use crate::seeding::randstrobes_query;
use crate::seeding::ry_equal;

/// Non-overlapping approximate match
#[derive(Clone, Debug)]
pub struct Nam {
    pub nam_id: usize,
    pub ref_start: usize,
    pub ref_end: usize,
    pub query_start: usize,
    pub query_end: usize,
    pub n_matches: usize,
    pub matching_bases: usize,
    pub ref_id: usize,
    pub score: f32,
    pub is_revcomp: bool,
    pub anchors: Vec<Anchor>,
}

impl Nam {
    pub fn ref_span(&self) -> usize {
        self.ref_end - self.ref_start
    }

    pub fn query_span(&self) -> usize {
        self.query_end - self.query_start
    }

    pub fn projected_ref_start(&self) -> usize {
        self.ref_start.saturating_sub(self.query_start)
    }

    /// Returns whether a NAM represents a consistent match between read and
    /// reference by comparing the nucleotide sequences of the first and last
    /// strobe (taking orientation into account).
    pub fn is_consistent(
        &self,
        read: &Read,
        references: &[RefSequence],
        k: usize,
        adna_mode: bool,
        ry_len: usize,
    ) -> bool {
        let ref_start_kmer = &references[self.ref_id].sequence[self.ref_start..self.ref_start + k];
        let ref_end_kmer = &references[self.ref_id].sequence[self.ref_end - k..self.ref_end];

        let seq = if self.is_revcomp {
            read.rc()
        } else {
            read.seq()
        };
        let read_start_kmer = &seq[self.query_start..self.query_start + k];
        let read_end_kmer = &seq[self.query_end - k..self.query_end];

        if adna_mode {
            ry_equal(ref_start_kmer, read_start_kmer, ry_len)
                && ry_equal(ref_end_kmer, read_end_kmer, ry_len)
        } else {
            ref_start_kmer == read_start_kmer && ref_end_kmer == read_end_kmer
        }
    }
}

impl Display for Nam {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Nam(ref_id={}, query: {}..{}, ref: {}..{}, rc={}, score={})",
            self.ref_id,
            self.query_start,
            self.query_end,
            self.ref_start,
            self.ref_end,
            self.is_revcomp as u8,
            self.score
        )?;
        Ok(())
    }
}

/// Determine whether the NAM represents a match to the forward or
/// reverse-complemented sequence by checking in which orientation the
/// first and last strobe in the NAM match
///
/// - If first and last strobe match in forward orientation, return true.
/// - If first and last strobe match in reverse orientation, update the NAM
///   in place and return true.
/// - If first and last strobe do not match consistently, return false.
pub fn reverse_nam_if_needed(
    nam: &mut Nam,
    read: &Read,
    references: &[RefSequence],
    k: usize,
    adna_mode: bool,
    ry_len: usize,
) -> bool {
    let ref_start_kmer = &references[nam.ref_id].sequence[nam.ref_start..nam.ref_start + k];
    let ref_end_kmer = &references[nam.ref_id].sequence[nam.ref_end - k..nam.ref_end];

    let kmers_match = |a: &[u8], b: &[u8]| -> bool {
        if adna_mode { ry_equal(a, b, ry_len) } else { a == b }
    };

    let (seq, seq_rc) = if nam.is_revcomp {
        (read.rc(), read.seq())
    } else {
        (read.seq(), read.rc())
    };
    let read_start_kmer = &seq[nam.query_start..nam.query_start + k];
    let read_end_kmer = &seq[nam.query_end - k..nam.query_end];
    if kmers_match(ref_start_kmer, read_start_kmer)
        && kmers_match(ref_end_kmer, read_end_kmer)
    {
        return true;
    }

    // False forward or false reverse (possible due to symmetrical hash values)
    // we need two extra checks for this - hopefully this will remove all the false matches we see
    // (true hash collisions should be very few)
    let read_len = read.len();
    let q_start_tmp = read_len - nam.query_end;
    let q_end_tmp = read_len - nam.query_start;
    // false reverse match, change coordinates in nam to forward
    let read_start_kmer = &seq_rc[q_start_tmp..q_start_tmp + k];
    let read_end_kmer = &seq_rc[q_end_tmp - k..q_end_tmp];
    if kmers_match(ref_start_kmer, read_start_kmer)
        && kmers_match(ref_end_kmer, read_end_kmer)
    {
        nam.is_revcomp = !nam.is_revcomp;
        nam.query_start = q_start_tmp;
        nam.query_end = q_end_tmp;
        true
    } else {
        false
    }
}

/// Obtain NAMs for a sequence record, doing rescue if needed.
///
/// NAMs are returned sorted by decreasing score
pub fn get_nams_by_chaining(
    sequence: &[u8],
    index: &StrobemerIndex,
    chainer: &Chainer,
    rescue_distance: usize,
    mcs_strategy: McsStrategy,
    rng: &mut Rng,
) -> (NamDetails, Vec<Nam>) {
    let timer = Instant::now();
    let query_randstrobes = randstrobes_query(sequence, &index.parameters);
    let time_randstrobes = timer.elapsed().as_secs_f64();

    trace!(
        "we have {} + {} randstrobes",
        query_randstrobes[0].len(),
        query_randstrobes[1].len()
    );

    let (mut nam_details, mut nams) =
        chainer.get_chains(&query_randstrobes, index, rescue_distance, mcs_strategy);

    let timer = Instant::now();

    nams.sort_by(|a, b| b.score.total_cmp(&a.score));
    shuffle_top_nams(&mut nams, rng);
    nam_details.time_sort_nams = timer.elapsed().as_secs_f64();
    nam_details.time_randstrobes = time_randstrobes;

    if log::log_enabled!(Trace) {
        trace!("Found {} NAMs", nams.len());
        let mut printed = 0;
        for nam in &nams {
            if nam.n_matches > 1 || printed < 10 {
                trace!("- {}", nam);
                printed += 1;
            }
        }
        if printed < nams.len() {
            trace!("+ {} single-anchor chains", nams.len() - printed);
        }
    }

    (nam_details, nams)
}

/// Shuffle the top-scoring NAMs. Input must be sorted by score.
/// This helps to ensure we pick a random location in case there are multiple
/// equally good ones.
fn shuffle_top_nams(nams: &mut [Nam], rng: &mut Rng) {
    if let Some(best) = nams.first() {
        let best_score = best.score;

        let pos = nams.iter().position(|nam| nam.score != best_score);
        let end = pos.unwrap_or(nams.len());
        if end > 1 {
            rng.shuffle(&mut nams[0..end]);
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::chainer::{Chainer, ChainingParameters};
    use crate::index::StrobemerIndex;
    use crate::io::fasta::RefSequence;
    use crate::seeding::{DEFAULT_AUX_LEN, SeedingParameters};

    #[test]
    fn test_partial_rymer_consistency() {
        // Build a reference containing a known k-mer whose canonical form
        // is the reverse complement. Construct a 200bp reference with unique flanking sequence so
        // the k-mer at the target position is the only hit.
        let target_kmer = b"TTGGCACAACTTCTTCTACAT";
        let k = 21;
        let ry_len = 8;

        let left_flank = b"ACGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATC";
        let right_flank = b"TAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTAGCTA";

        let mut ref_seq = Vec::new();
        ref_seq.extend_from_slice(left_flank);
        ref_seq.extend_from_slice(target_kmer);
        ref_seq.extend_from_slice(right_flank);
        let target_pos = left_flank.len();

        let references = vec![RefSequence {
            name: "test".to_string(),
            sequence: ref_seq.clone(),
        }];

        let parameters = SeedingParameters::from_read_length(
            40,
            Some(k),
            None,
            None,
            None,
            None,
            None,
            DEFAULT_AUX_LEN,
            true,
            Some(ry_len),
        )
        .unwrap();
        assert_eq!(parameters.syncmer.k, k);
        assert_eq!(parameters.ry_len, ry_len);

        let mut index = StrobemerIndex::new(&references, parameters.clone(), None);
        index.populate(0.0000001, 1);

        let chainer = Chainer::new(k, ChainingParameters::default());

        // Create a read that matches the target position but with C→T at
        // position 15 of the k-mer (a deamination-like change).
        // We include some flanking context from the reference so the read
        // is longer than k and produces multiple seeds.
        let deam_pos = 15; // within the target k-mer
        let read_start = target_pos;
        let read_len = 40;
        let mut read_seq: Vec<u8> = ref_seq[read_start..read_start + read_len].to_vec();
        assert_eq!(read_seq[deam_pos], b'C');
        read_seq[deam_pos] = b'T'; // C→T deamination

        let read = Read::new(&read_seq);

        let mut rng = Rng::with_seed(42);
        let (_details, nams) = get_nams_by_chaining(
            &read_seq,
            &index,
            &chainer,
            100,
            McsStrategy::Always,
            &mut rng,
        );

        // Find NAMs that overlap the correct reference position
        let correct_nams: Vec<&Nam> = nams
            .iter()
            .filter(|nam| {
                nam.ref_id == 0
                    && nam.ref_start <= target_pos + deam_pos
                    && nam.ref_end >= target_pos
            })
            .collect();

        assert!(
            !correct_nams.is_empty(),
            "Expected at least one NAM overlapping the correct reference position {}",
            target_pos,
        );

        // Check consistency of matches that contain at least one reverse complement
        for nam in &correct_nams {
            let consistent = nam.is_consistent(
                &read,
                &references,
                k,
                true,
                ry_len,
            );
            if nam.ref_start == left_flank.len() {            
                assert!(consistent, "Found inconsistent NAM {} for ry_len {}", nam, ry_len);
            }
        }
    }
}

File: cpp/nam.cpp
#include "nam.hpp"
#include "timer.hpp"
#include "hits.hpp"

bool operator==(const Match& lhs, const Match& rhs) {
    return (lhs.query_start == rhs.query_start) && (lhs.query_end == rhs.query_end) && (lhs.ref_start == rhs.ref_start) && (lhs.ref_end == rhs.ref_end);
}

namespace {


inline void add_to_matches_map_full(
    robin_hood::unordered_map<unsigned int, std::vector<Match>>& matches_map,
    int query_start,
    int query_end,
    const StrobemerIndex& index,
    size_t position
) {
    int min_diff = std::numeric_limits<int>::max();
    for (const auto hash = index.get_hash(position); index.get_hash(position) == hash; ++position) {
        int ref_start = index.get_strobe1_position(position);
        int ref_end = ref_start + index.strobe2_offset(position) + index.k();
        int diff = std::abs((query_end - query_start) - (ref_end - ref_start));
        if (diff <= min_diff) {
            matches_map[index.reference_index(position)].push_back(
                Match{query_start, query_end, ref_start, ref_end}
            );
            min_diff = diff;
        }
    }
}

/*
This function often produces the same Match multiple times.
This happens when multiple randstrobes involve a syncmer that has a low hash
value (and which therefore gets chosen as main hash multiple times).

For example, assume we have syncmers starting at positions 14, 21, 39, 51
and randstrobes like this:

14, 21, 39, 51
14------39
    21--39
        39--51

If 39 now has the lowest hash value, it will be chosen as the main hash for all
three randstrobes. Thus, when looking up that main hash, all three randstrobes
are found. After the start coordinate of the randstrobe is converted to the
start coordinate of the single syncmer (using strobe_extent_partial()), they all
result in a hit that starts at 39 (and has length k).
*/
inline void add_to_matches_map_partial(
    robin_hood::unordered_map<unsigned int, std::vector<Match>>& matches_map,
    int query_start,
    int query_end,
    const StrobemerIndex& index,
    size_t position
) {
    for (const auto hash = index.get_main_hash(position);
         index.get_main_hash(position) == hash;
         ++position
    ) {
        auto [ref_start, ref_end] = index.strobe_extent_partial(position);
        matches_map[index.reference_index(position)].push_back(
            Match{query_start, query_end, ref_start, ref_end}
        );
    }
}

} // namespace

void merge_matches_into_nams(
    robin_hood::unordered_map<unsigned int, std::vector<Match>>& matches_map,
    int k,
    bool sort,
    bool is_revcomp,
    std::vector<Nam>& nams  // inout
) {
    for (auto &[ref_id, matches] : matches_map) {
        if (sort) {
            std::sort(matches.begin(), matches.end(), [](const Match& a, const Match& b) -> bool {
                    // first sort on query starts, then on reference starts, finally prefer full matches over partial
                    return (a.query_start < b.query_start) ||
                            ( (a.query_start == b.query_start) && (a.ref_start < b.ref_start)) ||
                            ( (a.query_start == b.query_start) && (a.ref_start == b.ref_start) && (a.query_end > b.query_end)  );
                }
            );
        }

        std::vector<Nam> open_nams;
        int prev_q_start = 0;
        auto prev_match = Match{0,0,0,0};
        for (auto &m : matches) {
            if ( (prev_match == m) || ( ((m.query_end - m.query_start) == k) && (prev_match.query_start == m.query_start) && (prev_match.ref_start == m.ref_start)) )  { // is a redundant partial match
                continue;
            }
            bool is_added = false;
            for (auto & o : open_nams) {

                // Extend NAM
                if ((o.query_prev_match_startpos <= m.query_start) && (m.query_start <= o.query_end ) && (o.ref_prev_match_startpos <= m.ref_start) && (m.ref_start <= o.ref_end) ){
                    if ( (m.query_end > o.query_end) && (m.ref_end > o.ref_end) ) {
                        o.query_end = m.query_end;
                        o.ref_end = m.ref_end;
//                        o.previous_query_start = h.query_s;
//                        o.previous_ref_start = h.ref_s; // keeping track so that we don't . Can be caused by interleaved repeats.
                        o.query_prev_match_startpos = m.query_start;
                        o.ref_prev_match_startpos = m.ref_start;
                        o.n_matches ++;
//                        o.score += (float)1/ (float)h.count;
                        is_added = true;
                        break;
                    }
                    else if ((m.query_end <= o.query_end) && (m.ref_end <= o.ref_end)) {
//                        o.previous_query_start = h.query_s;
//                        o.previous_ref_start = h.ref_s; // keeping track so that we don't . Can be caused by interleaved repeats.
                        o.query_prev_match_startpos = m.query_start;
                        o.ref_prev_match_startpos = m.ref_start;
                        o.n_matches ++;
//                        o.score += (float)1/ (float)h.count;
                        is_added = true;
                        break;
                    }
                }

            }
            // Add to open matches
            if (!is_added){
                Nam n;
                n.query_start = m.query_start;
                n.query_end = m.query_end;
                n.ref_start = m.ref_start;
                n.ref_end = m.ref_end;
                n.ref_id = ref_id;
//                n.previous_query_start = h.query_s;
//                n.previous_ref_start = h.ref_s;
                n.query_prev_match_startpos = m.query_start;
                n.ref_prev_match_startpos = m.ref_start;
                n.n_matches = 1;
                n.is_revcomp = is_revcomp;
//                n.score += (float)1 / (float)h.count;
                open_nams.push_back(n);
            }

            // Only filter if we have advanced at least k nucleotides
            if (m.query_start > prev_q_start + k) {

                // Output all NAMs from open_matches to final_nams that the current match have passed
                for (auto &n : open_nams) {
                    if (n.query_end < m.query_start) {
                        int n_max_span = std::max(n.query_span(), n.ref_span());
                        int n_min_span = std::min(n.query_span(), n.ref_span());
                        float n_score;
                        n_score = ( 2*n_min_span -  n_max_span) > 0 ? (float) (n.n_matches * ( 2*n_min_span -  n_max_span) ) : 1;   // this is really just n_matches * ( min_span - (offset_in_span) ) );
//                        n_score = n.n_hits * n.query_span();
                        n.score = n_score;
                        n.nam_id = nams.size();
                        nams.push_back(n);
                    }
                }

                // Remove all NAMs from open_matches that the current match have passed
                auto c = m.query_start;
                auto predicate = [c](decltype(open_nams)::value_type const &nam) { return nam.query_end < c; };
                open_nams.erase(std::remove_if(open_nams.begin(), open_nams.end(), predicate), open_nams.end());
                prev_q_start = m.query_start;
            }
            prev_match = m;
        }

        // Add all current open_matches to final NAMs
        for (auto &n : open_nams) {
            int n_max_span = std::max(n.query_span(), n.ref_span());
            int n_min_span = std::min(n.query_span(), n.ref_span());
            float n_score;
            n_score = ( 2*n_min_span -  n_max_span) > 0 ? (float) (n.n_matches * ( 2*n_min_span -  n_max_span) ) : 1;   // this is really just n_matches * ( min_span - (offset_in_span) ) );
//            n_score = n.n_matches * n.query_span();
            n.score = n_score;
            n.nam_id = nams.size();
            nams.push_back(n);
        }
    }
}

robin_hood::unordered_map<unsigned int, std::vector<Match>> hits_to_matches(
    const std::vector<Hit>& hits,
    const StrobemerIndex& index
) {
    robin_hood::unordered_map<unsigned int, std::vector<Match>> matches_map;
    matches_map.reserve(100);

    for (const auto& hit : hits) {
        if (hit.is_filtered) {
            continue;
        }
        if (hit.is_partial) {
            add_to_matches_map_partial(matches_map, hit.query_start, hit.query_end, index, hit.position);
        } else {
            add_to_matches_map_full(matches_map, hit.query_start, hit.query_end, index, hit.position);
        }
    }

    return matches_map;
}

/*
 * Obtain NAMs for a sequence record, doing rescue if needed.
 * Return NAMs sorted by decreasing score.
 */
std::vector<Nam> get_nams(
    const std::array<std::vector<QueryRandstrobe>, 2>& query_randstrobes,
    const StrobemerIndex& index,
    AlignmentStatistics& statistics,
    Details& details,
    const MappingParameters &map_param
) {
    // Find hits
    Timer hits_timer;

    bool sorting_needed = false;
    std::array<std::vector<Hit>, 2> hits;
    for (int is_revcomp : {0, 1}) {
        bool sorting_needed1;
        HitsDetails hits_details1;
        std::tie(hits_details1, sorting_needed1, hits[is_revcomp]) = find_hits(query_randstrobes[is_revcomp], index, map_param.mcs_strategy, map_param.rescue_threshold);
        sorting_needed = sorting_needed || sorting_needed1;
        details.hits += hits_details1;
    }
    statistics.time_hit_finding += hits_timer.duration();

    std::vector<Nam> nams;

    // Rescue if requested and needed
    Timer chaining_timer;
    for (size_t is_revcomp = 0; is_revcomp < 2; ++is_revcomp) {
        auto matches_map = hits_to_matches(hits[is_revcomp], index);
        merge_matches_into_nams(matches_map, index.k(), sorting_needed, is_revcomp, nams);
    }
    details.nams = nams.size();
    statistics.time_chaining += chaining_timer.duration();

    return nams;
}

std::ostream& operator<<(std::ostream& os, const Hit& hit) {
    os << "Hit(query_start=" << hit.query_start << ", query_end=" << hit.query_end << ", position=" << hit.position << ", is_partial=" << hit.is_partial << ")";
    return os;
}

std::ostream& operator<<(std::ostream& os, const Match& match) {
    os << "Match(query_start=" << match.query_start << ", query_end=" << match.query_end << ", ref_start=" << match.ref_start << ", ref_end=" << match.ref_end << ")";
    return os;
}

std::ostream& operator<<(std::ostream& os, const Nam& n) {
    os << "Nam(ref_id=" << n.ref_id << ", query: " << n.query_start << ".." << n.query_end << ", ref: " << n.ref_start << ".." << n.ref_end << ", rc=" << static_cast<int>(n.is_revcomp) << ", n_matches=" << n.n_matches << ", score=" << n.score << ")";
    return os;
}
